include/amber/amber_vulkan.h

```diff
@@ -15,16 +15,43 @@
 #ifndef AMBER_AMBER_VULKAN_H_
 #define AMBER_AMBER_VULKAN_H_
 
+#include <string>
+#include <vector>
+
+#include "amber/amber.h"
 #include "vulkan/vulkan.h"
 
 namespace amber {
 
 /// Configuration for the Vulkan Engine.
 struct VulkanEngineConfig : public EngineConfig {
-  /// The VkDevice to use for the tests.
-  VkDevice device;
+  /// The VkPhysicalDevice to be used for the tests.
+  VkPhysicalDevice physical_device = VK_NULL_HANDLE;
+
+  /// Physical device features available for |physical_device|.
+  VkPhysicalDeviceFeatures available_features = {};
+
+  /// Physical device extensions available for |physical_device|.
+  std::vector<std::string> available_extensions;
+
+  /// The given queue family index to be used for the tests.
+  uint32_t queue_family_index;
+
+  /// The VkQueue to be used for the tests.
+  VkQueue queue;
+
+  /// The VkDevice to be used for the tests.
+  VkDevice device = VK_NULL_HANDLE;
 };
 
+/// Returns required physical device features from |recipes|.
+VkPhysicalDeviceFeatures GetRequiredVulkanFeatures(
```

Comment:
These should be methods on Recipe themselves.

std::vector<std::string> Recipe::GetRequiredFeatures()
std::vector<std::string> Recipe::GetRequiredExtensions()

Reply:
We need `VkPhysicalDeviceFeatures` which is not `std::vector<std::string>`. It is Vulkan specific object.
In addition, for WebGPU, we might need other method.

We must add `#if AMBER_ENGINE_VULKAN` when we use `VkPhysicalDeviceFeatures`.
I intended to put all vulkan related APIs in include/amber/amber_vulkan.h.

An alternative is to define a new structure `EngineFeature` and use `struct VulkanEngineFeature : public EngineFeature { VkPhysicalDeviceFeatures features; }` and add `std::unique<EngineFeature> Recipe::GetRequiredFeatures()` to `Recipe`.

Reply:
We can convert to a VkPhysicalDevice inside the amber code, same as is done currently. The Recipe should return strings I think.

In amber_vulkan we will need to pass VkPhysicalDeviceFeatures if that's what CTS provides, but that's already contained in vulkan specific code and doesn't need an AMBER_ENGINE_VULKAN ifdef.

Reply:
(The reason for returning strings instead of Vk items is that we'd have to implement the recipe code twice to do a Dawn equivalent. Returning strings will let us re-use the same logic for multiple backends)